samples/common/wifi.c

```diff
@@ -1,44 +1,143 @@
+	WIFI_STATE_IDLE,
+	WIFI_STATE_WAITING_FOR_LOWER_UP,
+	WIFI_STATE_CONNECTING,
+	WIFI_STATE_DISCONNECTING,
+	WIFI_STATE_READY,
+	WIFI_STATE_LAST = WIFI_STATE_READY,
+	WIFI_STATE_INVALID,
+	WIFI_STATE_WAIT,
+};
+
+enum wifi_event {
+	WIFI_EVENT_START,
+	WIFI_EVENT_LOWER_UP,
+	WIFI_EVENT_CONNECTED,
+	WIFI_EVENT_DISCONNECTED,
+	WIFI_EVENT_IP_ADD,
+	WIFI_EVENT_IP_DEL,
+};
+
+struct wifi_manager_config {
+	k_thread_stack_t *workq_stack;
+	size_t workq_stack_size;
+	struct k_work_queue_config workq_config;
+};
+
+enum wifi_manager_flags {
+	WIFI_FLAG_ENABLE,
+	WIFI_FLAG_RECONNECT,
+};
+
+struct wifi_manager_data {
+	struct net_if *iface;
+
+	struct k_work_q workq;
+	enum wifi_state state;
+
+	atomic_t net_state;
+	struct k_work net_mgmt_event_work;
+	int net_state_old;
+
+	atomic_t flags;
+
+	enum wifi_event event;
+	struct k_work event_work;
+
+	struct net_mgmt_event_callback ipv4_mgmt_cb;
 	struct net_mgmt_event_callback wifi_mgmt_cb;
-	struct k_sem connect_sem;
-	int connect_status;
 };
 
-static void wifi_mgmt_event_handler(struct net_mgmt_event_callback *cb,
-				    uint32_t mgmt_event, struct net_if *iface)
+static const char *wifi_state_str(enum wifi_state state)
 {
-	struct wifi_data *wifi = CONTAINER_OF(cb, struct wifi_data, wifi_mgmt_cb);
-	const struct wifi_status *status = cb->info;
-
-	switch (mgmt_event) {
-	case NET_EVENT_WIFI_CONNECT_RESULT:
-		LOG_INF("%s with status: %d", "Connected", status->status);
-		wifi->connect_status = status->status;
-		k_sem_give(&wifi->connect_sem);
-		break;
-	case NET_EVENT_WIFI_DISCONNECT_RESULT:
-		LOG_INF("%s with status: %d", "Disconnected", status->status);
-		wifi->connect_status = -EFAULT;
-		k_sem_give(&wifi->connect_sem);
-		break;
+	switch (state) {
+	case WIFI_STATE_IDLE:
+		return "IDLE";
+	case WIFI_STATE_WAITING_FOR_LOWER_UP:
+		return "WAIT_FOR_LOWER_UP";
+	case WIFI_STATE_CONNECTING:
+		return "CONNECTING";
+	case WIFI_STATE_DISCONNECTING:
+		return "DISCONNECTING";
+	case WIFI_STATE_READY:
+		return "READY";
+	case WIFI_STATE_INVALID:
+		return "INVALID";
+	case WIFI_STATE_WAIT:
 	default:
-		break;
+		return "WAIT";
 	}
+
+	return "";
 }
 
+static const char *wifi_event_str(enum wifi_event event)
+{
+	switch (event) {
+	case WIFI_EVENT_START:
+		return "START";
+	case WIFI_EVENT_LOWER_UP:
+		return "LOWER_UP";
+	case WIFI_EVENT_CONNECTED:
+		return "CONNECTED";
+	case WIFI_EVENT_DISCONNECTED:
+		return "DISCONNECTED";
+	case WIFI_EVENT_IP_ADD:
+		return "IP_ADD";
+	case WIFI_EVENT_IP_DEL:
+		return "IP_DEL";
+	}
+
+	return "";
+}
+
+#define MAP_OFFSET	1
+
+#define IDL		(WIFI_STATE_IDLE + MAP_OFFSET)
+#define LUP		(WIFI_STATE_WAITING_FOR_LOWER_UP + MAP_OFFSET)
+#define CON		(WIFI_STATE_CONNECTING + MAP_OFFSET)
+#define DIS		(WIFI_STATE_DISCONNECTING + MAP_OFFSET)
+#define RDY		(WIFI_STATE_READY + MAP_OFFSET)
+#define INV		(WIFI_STATE_INVALID + MAP_OFFSET)
+#define WAIT(ms)	(WIFI_STATE_WAIT + MAP_OFFSET + (ms))
+
+static const int wifi_state_change_map[][WIFI_STATE_LAST + 1] = {
```

Comment:
Why not use the [State Machine Framework](https://docs.zephyrproject.org/latest/services/smf/index.html)?

Reply:
I looked briefly at that framework months ago and... I was just expecting something more I guess. With a hierarchical state machine this might be a good thing to use, but for flat state machine I got the impression that it didn't provide that much tools to speed up development or reuse.

I guess that the real answer could be provided only by trying to reimplement what we have in this PR with State Machine Framework and see how it looks in practice. This approach should be faster and give more precise answer than theoretical considerations.